File: AI-Reply-Android/app/src/main/java/kz/yerek/aireply/domain/model/UserProfile.kt
package kz.yerek.aireply.domain.model

import kotlinx.serialization.Serializable
import kz.yerek.aireply.core.text.clampToCodePoints

/**
 * The lightweight personal communication profile built during onboarding.
 *
 * It never leaves the device except as part of an AI request the user
 * explicitly triggered, and even then only the fields the model actually needs
 * are sent.
 *
 * Every field carries a default, which is how a profile written by an older
 * build still loads after an update instead of resetting what the user typed.
 */
@Serializable
data class UserProfile(
    /**
     * Free text: who the user is and how they usually communicate. Typed or
     * dictated. This is the "About me" answer.
     */
    val descriptionText: String = "",

    /**
     * What the user does for a living, in their words: "online clothing store
     * owner", "интернет-маркетолог", "дизайнер".
     */
    val role: String = "",

    /**
     * What the user provides, and the rules that hold across every
     * conversation. Per-template context is layered on top of this, never
     * instead of it.
     */
    val business: BusinessContext = BusinessContext.EMPTY,

    /** Default register, used when a template does not override it. */
    val preferredTone: ReplyTone = ReplyTone.NATURAL,

    /**
     * Which conversation kinds the user said they care about. Used to preselect
     * which templates appear in the keyboard bar, never sent to the model.
     */
    val activeRelationships: Set<RelationshipKind> = RelationshipKind.builtIns.toSet(),

    val workingHours: WorkingHours = WorkingHours.DEFAULT,

    val hasCompletedOnboarding: Boolean = false
) {

    fun withRole(value: String) = copy(role = value.clampToCodePoints(MAX_ROLE))

    fun withDescription(value: String) =
        copy(descriptionText = value.clampToCodePoints(MAX_DESCRIPTION))

    /** What actually goes in an AI request. Trimmed, never the whole object. */
    val promptDescription: String get() = descriptionText.trim()

    /**
     * True once the user has told us anything at all about themselves. Used to
     * decide whether the home screen should still be nudging them to.
     */
    val hasAnyContext: Boolean
        get() = promptDescription.isNotEmpty() || role.isNotBlank() || !business.isEmpty

    companion object {
        /**
         * Maximum stored length for the free-text answer. The product asks for
         * roughly 500-1000 characters; 1000 is the cap and the editor shows a
         * live counter well before it.
         */
        const val MAX_DESCRIPTION = 1000

        /**
         * Short single-line answer. A role is "online clothing store owner",
         * not an essay, and a cap that small keeps the per-request prompt cheap.
         */
        const val MAX_ROLE = 120

        val EMPTY = UserProfile()
    }
}

File: AI-Reply/Shared/Model/UserProfile.swift
import Foundation

/// The lightweight personal communication profile built during onboarding.
///
/// It never leaves the device except as part of an AI request the user
/// explicitly triggered, and only the two fields the model actually needs
/// (`description`, `preferredTone`) are sent even then.
struct UserProfile: Codable, Hashable, Sendable {

    /// Maximum stored length. The brief asks for roughly 500-1000 characters;
    /// 1000 is the cap and the editor shows a live counter well before it.
    static let maximumDescriptionCharacters = 1000

    /// Short single-line answers. A role is "online clothing store owner", not
    /// an essay, and a cap that small keeps the per-request prompt cheap.
    static let maximumRoleCharacters = 120

    /// Free text: who the user is and how they usually communicate. Typed or
    /// dictated. This is the "About me" answer.
    var descriptionText: String

    /// What the user does for a living, in their words: "online clothing store
    /// owner", "интернет-маркетолог", "дизайнер".
    var role: String

    /// What the user provides, and the rules that hold across every
    /// conversation. Per-template context is layered on top of this, never
    /// instead of it.
    var business: BusinessContext

    /// Default register, used when a template does not override it.
    var preferredTone: ReplyTone

    /// Which conversation kinds the user said they care about. Used to preselect
    /// which templates appear in the keyboard bar, never sent to the model.
    var activeRelationships: Set<RelationshipKind>

    var workingHours: WorkingHours

    var hasCompletedOnboarding: Bool

    static let empty = UserProfile(
        descriptionText: "",
        role: "",
        business: .empty,
        preferredTone: .natural,
        activeRelationships: Set(RelationshipKind.builtIns),
        workingHours: .default,
        hasCompletedOnboarding: false
    )

    mutating func setRole(_ value: String) {
        role = value.unicodeScalars.count <= Self.maximumRoleCharacters
            ? value
            : String(value.prefix(Self.maximumRoleCharacters))
    }

    /// Clamped by scalar count, matching the backend's own counting so the two
    /// limits mean the same thing.
    mutating func setDescription(_ value: String) {
        let trimmed = value
        if trimmed.unicodeScalars.count <= Self.maximumDescriptionCharacters {
            descriptionText = trimmed
        } else {
            descriptionText = String(trimmed.prefix(Self.maximumDescriptionCharacters))
        }
    }

    /// What actually goes in an AI request. Trimmed, never the whole struct.
    var promptDescription: String {
        descriptionText.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // Decoding tolerates a profile written by an older build that had fewer
    // fields, so an app update never wipes what the user typed.
    /// True once the user has told us anything at all about themselves. Used
    /// to decide whether the home screen should still be nudging them to.
    var hasAnyContext: Bool {
        !promptDescription.isEmpty
            || !role.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            || !business.isEmpty
    }

    enum CodingKeys: String, CodingKey {
        case descriptionText
        case role
        case business
        case preferredTone
        case activeRelationships
        case workingHours
        case hasCompletedOnboarding
    }

    init(
        descriptionText: String,
        role: String = "",
        business: BusinessContext = .empty,
        preferredTone: ReplyTone,
        activeRelationships: Set<RelationshipKind>,
        workingHours: WorkingHours,
        hasCompletedOnboarding: Bool
    ) {
        self.descriptionText = descriptionText
        self.role = role
        self.business = business
        self.preferredTone = preferredTone
        self.activeRelationships = activeRelationships
        self.workingHours = workingHours
        self.hasCompletedOnboarding = hasCompletedOnboarding
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        descriptionText = try container.decodeIfPresent(String.self, forKey: .descriptionText) ?? ""
        role = try container.decodeIfPresent(String.self, forKey: .role) ?? ""
        business = try container.decodeIfPresent(BusinessContext.self, forKey: .business) ?? .empty
        preferredTone = try container.decodeIfPresent(ReplyTone.self, forKey: .preferredTone) ?? .natural
        activeRelationships = try container.decodeIfPresent(Set<RelationshipKind>.self, forKey: .activeRelationships)
            ?? Set(RelationshipKind.builtIns)
        workingHours = try container.decodeIfPresent(WorkingHours.self, forKey: .workingHours) ?? .default
        hasCompletedOnboarding = try container.decodeIfPresent(Bool.self, forKey: .hasCompletedOnboarding) ?? false
    }
}
